src/qrcode.js

```diff
@@ -0,0 +1,330 @@
+    const quietModuleCount = vqr.getModuleCount() + 2 * quiet;
+
+    function isDark(row, col) {
+        row -= quiet;
+        col -= quiet;
+
+        if (row < 0 || row >= qrModuleCount || col < 0 || col >= qrModuleCount) {
+            return false;
+        }
+        return vqr.isDark(row, col);
+    }
+
+    function addBlank(l, t, r, b) {
+        const prevIsDark = qr.isDark;
+        const moduleSize = 1 / quietModuleCount;
+
+        qr.isDark = function (row, col) {
+            const ml = col * moduleSize;
+            const mt = row * moduleSize;
+            const mr = ml + moduleSize;
+            const mb = mt + moduleSize;
+
+            return prevIsDark(row, col) && (l > mr || ml > r || t > mb || mt > b);
+        };
+    }
+
+    qr.text = text;
+    qr.level = level;
+    qr.version = version;
+    qr.moduleCount = quietModuleCount;
+    qr.isDark = isDark;
+    qr.addBlank = addBlank;
+
+    return qr;
+}
+
+function createMinQRCode(text, level, minVersion, maxVersion, quiet) {
+    minVersion = Math.max(1, minVersion || 1);
+    maxVersion = Math.min(40, maxVersion || 40);
+    for (let version = minVersion; version <= maxVersion; version += 1) {
+        try {
+            return createQRCode(text, level, version, quiet);
+        } catch (err) { /* empty */ }
+    }
+    return undefined;
+}
+
+function drawOnCanvas(element, options, imageElement = null) {
+    const settings = { ...options, image: imageElement };
+    console.log('[CHECKING ELEMENT IN DRAW CANVAS] : ', element, options, imageElement);
+
+    element.width = settings.size;
+    element.height = settings.size;
+
+    const qr = createMinQRCode(settings.text, settings.ecLevel, settings.minVersion, settings.maxVersion, settings.quiet);
+
+    if (!qr) {
+        return null;
+    }
+
+    const context = element.getContext('2d');
+
+    drawBackground(qr, context, settings);
+    drawModules(qr, context, settings);
+
+    return element;
+}
+
+function loadImage(element, options) {
+    const { image } = options;
+    const imageElement = new Image();
+
+    imageElement.addEventListener('load', () => drawOnCanvas(element, options, imageElement));
+    imageElement.addEventListener('error', () => {
+        throw new Error('Please provide a valid image');
+    });
+
+    if (image && typeof image === 'string') {
+        imageElement.src = image;
+    } else {
+        throw new Error('An Image is required for Mode 3 and Mode 4');
+    }
+}
+
+function createCanvas(element, options) {
+    if (options.mode === 3 || options.mode === 4) {
+        return loadImage(element, options);
+    }
+    return drawOnCanvas(element, options);
+}
+
+class qrcode {
+    constructor(ele) {
+        this.element = ele;
+    }
+
+    /**
+     * Returns a Promise resolved when done with the canvas element
+     * @returns {*}
+     */
+    generate(opts) {
+        if (!(this.element.tagName === 'DIV' || this.element.tagName === 'CANVAS')) {
```

Comment:
Better with 

```javascript
if ( !( this.element instanceof HTMLDivElement || this.element instanceof HTMLCanvasElement) )
```

Reply:
And what happen if the element has no property tagName (if it's undefined ?)

Reply:
You should move that test in constructor and throw TypeError instead

Reply:
Will `throw new TypeError(message)` be enough? Or should I do - `throw new TypeError(message, file, lineNumber)` as per MDN documentation?